### DISCUSSION TITLE
any solution to log request param and response data with api middleware?

### DISCUSSION THREAD

**Discussion by millievn** [vercel/next.js#30681] [Help] [answered]
The docs mention the logging but no example as expected.
How to log both log request param and response data with some logger like winston ?

**Comment by jamesmosier** [comment, marked as answer]
I believe something like this should work to get the request & response:

```js
import { NextResponse } from 'next/server'
import logger from 'your-winston-logger'

export default function middleware(request) {
  const response = NextResponse.next()
  logger.log(request);
  logger.log(response);
  return response;
}
```

**Reply by millievn** [reply]
So i guess NextResponse.next should be synchronous function. [The docs](https://github.com/senchalabs/connect) says `it can use Connect compatible middleware`.  How it works to make an asynchronous api handler ? Like koa or express or none of both?

**Reply by jamesmosier** [reply]
Oh sorry, so the code snippet I wrote above was for the new Middleware feature (https://nextjs.org/docs/middleware) which is when you create a `_middleware.js` file to handle all middleware for your pages/api routes (or a set of pages/api routes)

The [API middleware](https://nextjs.org/docs/api-routes/api-middlewares) can be used by doing something like:

```js
const withLogging = (handler) => {
  return async (req, res) => {
    logger.log(req)
    logger.log(res)
    return handler(req, res);
  };
};

// pages/api/my-api-route.js
function MyApiRoute(req, res) {
  // your api logic
  res.json({myResponse: 123})
}

export default withLogging(MyApiRoute)
```

**Reply by toto1384** [reply]
Is there any way I can get the response body? (it returns nothing when i run `await res.text()` or `await res.json()`)

**Reply by plrthink** [reply]
> Is there any way I can get the response body? (it returns nothing when i run `await res.text()` or `await res.json()`)

Have you solved the problem?

**Reply by andresg747** [reply]
Here's the explanation why the middleware don't produce a response body.
https://nextjs.org/docs/messages/middleware-upgrade-guide#no-response-body

**Comment by tomups** [comment]
You can patch the Next.js server to have full access to the request and response objects, and log them with pino-http for example:

https://www.tomups.com/posts/log-nextjs-request-response-as-json/
